File: Python3/623. Add One Row to Tree.py
class Solution:
    def addOneRow(self, root, v, d):
        """
        :type root: TreeNode
        :type v: int
        :type d: int
        :rtype: TreeNode
        """
        # if d == 1: # 广度搜索 64ms
        #     newroot = TreeNode(v)
        #     newroot.left = root
        #     return newroot
        # lastlevel = []
        # cur = 1
        # curlevel = [root]
        # nextlevel = []
        # while cur < d:
        #     for node in curlevel:
        #         if node:
        #             nextlevel.append(node.left)
        #             nextlevel.append(node.right)
        #     lastlevel = curlevel
        #     curlevel = nextlevel
        #     nextlevel = []
        #     cur += 1
        # count = 0
        # for node in lastlevel:
        #     if node:
        #         templeft, tempright = TreeNode(v), TreeNode(v)
        #         templeft.left = curlevel[count]
        #         tempright.right = curlevel[count+1]
        #         count += 2
        #         node.left = templeft
        #         node.right = tempright
        # return root
        
        if d == 1: # 队列解决 64ms
            newroot = TreeNode(v)
            newroot.left = root
            return newroot
        queue = [(root, 1)]
        for node, level in queue:
            if node.left:
                queue.append((node.left, level+1))
            if node.right:
                queue.append((node.right, level+1))
            if level == d-1:
                newleft = TreeNode(v)
                newleft.left = node.left
                newright = TreeNode(v)
                newright.right = node.right
                node.left = newleft
                node.right = newright
        return root

File: Golang/623. Add One Row to Tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type TreeNode2 struct { // 找不出错误
    Node *TreeNode
    Level int
}
func addOneRow(root *TreeNode, v int, d int) *TreeNode {
    if d == 1 {
        newroot := &TreeNode{Val: v, Left: root}
        return newroot
    }
    
    newroot := &TreeNode2{Node: root, Level: 1}
    
    queue := []*TreeNode2{newroot}
    
    for _, node := range queue {
        if node.Node.Left != nil {
            newleft := &TreeNode2{Node: node.Node.Left, Level: node.Level+1}
            queue = append(queue, newleft)
        }
        
        if node.Node.Right != nil {
            newright := &TreeNode2{Node: node.Node.Right, Level: node.Level+1}
            queue = append(queue, newright)
        }
        
        if node.Level == d-1 {
            newleft := &TreeNode{Val: v, Left: node.Node.Left}
            newright := &TreeNode{Val: v, Right: node.Node.Right}
            node.Node.Left = newleft
            node.Node.Right = newright
        }
    }
    return root
}
